refector : 실시간 매칭 리펙토링

ReadnThink opened the pull request:
# 리펙토링
- 리펙토링으로 코드량을 줄이고 클린 코드를 적용하여 가독성 증가
    - **약 50% 감소 (268줄 → 137줄)**

## 리펙토링을 하며 느낀점
처음 개발을 했던 코드를 보면서 그때의 열정을 다시 느낄수 있었습니다.
어떻게든 구현을 해야했고, 결국 구현을 했지만 지금 돌아보니 부족한점이 많았습니다.

두가지를 중점으로 리펙토링을 진행하였습니다.

1. 동료 개발자가 알아볼 수 있는 클린한 코드인가?
2. 효율적으로 더 개선할 곳은 없을까?

리펙토링을 진행하며 리펙토링이란 개발자의 성장을 위한 과정이라고 느껴졌습니다.
실시간 매칭기능의 첫 리펙토링 커밋을 시작으로 지속해서 안정적이고 유연하고 클린한 코드를 작성해야겠다는
생각이 들었습니다.    


# 기능 구현

💡[**LiveMatchService**](https://github.com/ReadnThink/pocoapoco-refectoring/blob/33fa9e943236fb6e1fb73b749da21f7dfea34fee/src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java)  최종 리펙토링된 코드입니다.

1. 매칭 취소로직 HashMap -> RedisSet으로 변경하여 Scale Out시에도 매칭을 취소할 수 있게 변경하였습니다. 
2. 개발 처음당시 작성했던 실시간 매칭 코드를 리펙토링 하였습니다.
3. 단순 반복문, 하드코딩, 너무 많은 메소드 인자, 명확하지 않은 변수명을 리펙토링 하였습니다.
4. 배열보다는 컬렉션을 사용하여 안전한 코드를 작성하였습니다.
5. Sse로직을 SseSender를 사용해 처리하였습니다. * 기입하기
   - Sse를 전송하는 로직이 자주 사용되어 SendSse함수를 만들어
    재사용하도록 하였습니다.
6. Crew를 만드는 코드가 길어 빌더패턴을 적용하지 않고 Static 메서드를 커스텀하여 Crew를 생성하였습니다.
   - 빌더패턴을 사용하는 이유는 로직이 변경되어도 객체 생성시 필드값을 자유롭게 설정하는 것이였는데 
   랜덤매칭시 Crew를 만들때는 인자가 제한적이고, 코드가 너무 길다고 판단하여 Static 메서드를 사용했습니다.


# 주목 포인트

1. 개발 처음 만든 실시간 매칭기능을 얼마나 가독성 있고 효율적으로 리펙토링 하였는가? 

[실시간 매칭 구현하기](https://velog.io/@readnthink/%EC%8B%A4%EC%8B%9C%EA%B0%84-%EB%A7%A4%EC%B9%AD%EC%9D%84-%EC%9C%84%ED%95%9C-Redis-%EA%B3%B5%EB%B6%80) 해당 글을 통해 전체적인 흐름을 간략하게 이해하실 수 있습니다!

## 불필요한 반복문들을 제거하였습니다.

iterator 코드
```java
// 대기열에 있는 User들의 name을 넣기위해 선언 -> redis에서 삭제, user 찾을때 사용
            String[] UserListInRedis = new String[4];

            // 대기열의 User들을 확인하기 위한 iterator
            Set<String> range = redisTemplate.opsForZSet().range(sport, 0, 3);

            Iterator<String> iterator = range.iterator();
            // iterator 돌면서 user string 으로 저장 -> user 찾기 위해
            int listCnt = 0;
            while (iterator.hasNext()) {
                String next = iterator.next();
                log.info("next = {}", next);
                UserListInRedis[listCnt] = next;
                log.info("redis의 대기열 = {}",UserListInRedis[listCnt]);
                listCnt++;
            }

            // User를 찾는다 -> 모임을 만들기 위해
            User fistUser = findUserFromRedis(UserListInRedis[0]);
            User secondUser = findUserFromRedis(UserListInRedis[1]);
            User thirdUser = findUserFromRedis(UserListInRedis[2]);


            Crew savedLiveMatchCrew = makeCrew(fistUser, secondUser, thirdUser, sport);
```

변경코드
```java

final List<User> users = findUser(redisTemplate.opsForZSet().range(sport, 0, CREW_LIMIT));
Crew savedLiveMatchCrew = crewRepository.save(Crew.makeRandomMatchingCrew(users, sport, CREW_LIMIT));

```
1. Builder 패턴을 사용하다 Static 메서드를 사용해 훨씬 더 깔끔하고 가독성이 코드로 변경이 가능하였습니다.
2. 기존에는 findUserFromRedis를 사용해 User를 한명씩찾아서 받아왔습니다.
하지만 이제는 findUser에 Redis가 반환하는 Set으로 변경하였습니다. 

<br/>

## 중복되는 로직을 메서드로 변경
중복되는 로직을 메서드로 변경하여 재사용 하였습니다.

변경 전
```java
Participation firstParticipation = Participation.builder()
                .status(2)
                .user(fistUser)
                .crew(savedLiveMatchCrew)
                .title(sport + "실시간 매칭🔥")
                .build();

        Participation secParticipation = Participation.builder()
                .status(2)
                .user(secondUser)
                .crew(savedLiveMatchCrew)
                .title(sport + "실시간 매칭🔥")
                .build();

        Participation thirdParticipation = Participation.builder()
                .status(2)
                .user(thirdUser)
                .crew(savedLiveMatchCrew)
                .title(sport + "실시간 매칭🔥")
                .build();
```

변경 후
```java
        Participation firstParticipation = makeParticipation(savedLiveMatchCrew, sport, users[0]);

       private static Participation makeParticipation(final Crew savedLiveMatchCrew, final String sport, final User firstUser) {
        return Participation.builder()
                .status(2)    // todo Participation status 상태 Enum으로 가독성 높이기
                .user(firstUser)
                .crew(savedLiveMatchCrew)
                .title(sport + "실시간 매칭🔥")
                .build();
    }
```

## Enum으로 변한  하드코딩
기존에 하드코딩되어있던 Sse에 전송될 name을 Enum으로 관리하여
한군데에서 관리할 수 있도록 하였습니다.
AlarmTypeEnum을 기준으로 보신다면 코드를 리뷰하시기 편하실 것입니다. 

# 추가적으로 더 필요한 곳

* Redis에 userName을 저장하여서 불필요한 로직들이 생기고 있습니다.
Redis에 User를 저장하여서 리펙토링 할 수 있을지 검토해봐야 합니다.

ReadnThink committed:
refector : 실시간 매칭 리펙토링

- 개발 처음당시 작성했던 실시간 매칭 코드를 리펙토링 하였습니다.
- 단순 반복문, 하드코딩, 너무 많은 메소드 인자, 명확하지 않은 변수명을
  리펙토링 하였습니다.

doxxx93 commented:
어떤 로직인지 설명 해주실 수 있을까요 ? 어떤 로직인지 context가 없어서 어떤 점을 봐야할지 잘 모르겠습니다.

ReadnThink commented:
> 어떤 로직인지 설명 해주실 수 있을까요 ? 어떤 로직인지 context가 없어서 어떤 점을 봐야할지 잘 모르겠습니다.

실시간 매칭을 하면 크루 방 생성, 채팅방 생성 등 여러 기능들이 있기에
모든 로직을 자세히 알고 싶으시다면 따로 디스코드에서 설명해드리겠습니다:)

다만 어느정도의 흐름을 읽는것이라면 [실시간 매칭 구현하기](https://velog.io/@readnthink/%EC%8B%A4%EC%8B%9C%EA%B0%84-%EB%A7%A4%EC%B9%AD%EC%9D%84-%EC%9C%84%ED%95%9C-Redis-%EA%B3%B5%EB%B6%80) 이 글을 통해 충분히 이해하실수 있을거같습니다!

만약 추가적인 설명이 필요하시다면 언제든 말씀해 주세요!

doxxx93 commented on src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java:
```diff
@@ -15,107 +14,94 @@
 
-        // redis에 대기열 순서대로 삽입, 현재 시간을 score로 잡음
-        redisTemplate.opsForZSet().add(sport, username, System.currentTimeMillis());
+        redisTemplate
```
현재 대기열을 위해서 sortedSet을 이용하고 계신데, 어떤 자료구조들 중에서 sortedSet을 결정하게 됐을까요?

여러 해결방법중 redis를 선택하게 된 이유가 뭘까요?

ReadnThink replied:
Queue, List, sortedSet중에서 sortedSet을 선택하게 되었습니다.

가장 큰 이유는 조회시 시간복잡도에 따른 문제였습니다.

sortedSet은 set의 자료구조이면서 score에 따라 순서를 보장할 수 있습니다.
즉, Queue의 방식으로 순서를 보장하여 sortedSet에서 꺼내고 중복확인을 O(1)의 시간복잡도로 해결할 수 있습니다.

반면 Queue와 List는 순서는 보장하지만 조회시 O(n)의 시간복잡도가 필요합니다.

다만, sortedSet이 정말 가장 효율적인것인가? 라는것은 조금더 고민해 보아야 합니다.
Redis의 pub/sub, stream 또는 카프카를 고려중이나 아직 자세히 공부한것은 아니기에
자세히 공부 후 리펙토링 예정입니다.

doxxx93 commented on src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java:
```diff
@@ -15,107 +14,94 @@
+                .add(
+                        sport,
+                        username,
+                        System.currentTimeMillis()
```
username이라는 String과 현재시간을 가중치로 넣는 것 같은데, redis에서 지원하는 가중치에 대한 처리방법이 있을지

또 

객체 자체를 저장하는 건 어떨까요?

ReadnThink replied:
객체를 저장하는 방법에 대해 고민중에 있습니다.
확실히 username을 저장하는것보다 User를 직접 저장해서 사용하는것이 
UserRepository를 사용해 DB에 접근하여 I/O 발생을 줄이고 Redis(Cache)를 알맞게 사용하는 방법인것 같습니다.

Redis의 Sorted Set에는 Score를 통해 순서를 정하는데 이떄 숫자를 통해 정할수 있습니다.
redis에서 지원하는 처리방법에 대해 더 공부해 보겠습니다.

doxxx93 commented on src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java:
```diff
@@ -15,107 +14,94 @@
-//         대기리스트에 3명이 들어왔다면
-        if (randomMatchListInRedis >= 3) {
+        // todo if문을 꼭 써야하나?
+        if (matchingUsersCnt >= CREW_LIMIT) {
```
랜덤 매칭이라는 메서드가 현재는 대기열에 등록 하는 것과 특정 인원이 모이면 매칭이 성사되는 역할 까지 갖고 있는 것 같아요. 맞을까요?

ReadnThink replied:
네 맞습니다.
다만 지금 구조는 if문을 사용해서 3개 이상일때만 매칭이 잡히는 구조입니다.

조금 더 탄탄한 코드를 만들기 위해 추가적인 리펙토링이 필요한 상태입니다.
Redis의 pub/sub 또는 카프카를 고려중이나 아직 자세히 공부한것은 아니기에
자세히 공부 후 리펙토링 예정입니다.

doxxx93 commented on src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java:
```diff
@@ -15,107 +14,94 @@
-
-
-            Crew savedLiveMatchCrew = makeCrew(fistUser, secondUser, thirdUser, sport);
+            final String[] userNames = usersToRandomMatch
```
usersToRandomMatch 를 중간에, temp 변수로 둬야하는 이유가 있을까요?

ReadnThink replied:
temp변수는 불필요한 상황입니다.
temp변수를 제거하고 아래의 코드로 변경하였습니다.

```java
final List<User> users = findUser(redisTemplate.opsForZSet().range(sport, 0, CREW_LIMIT));
```

doxxx93 commented on src/main/java/teamproject/pocoapoco/service/livematch/LiveMatchService.java:
```diff
@@ -15,107 +14,94 @@
+            final String[] userNames = usersToRandomMatch
+                    .stream()
+                    .toArray(String[]::new);
+            final User[] users = findUser(userNames);
```
배열보다는 컬렉션을 이용해보시는 건 어떨까요?

이펙티브 자바. 아이템28: 배열보다는 리스트를 사용하라 아이템을 한번 확인해보시면 좋을 것 같아요.

ReadnThink replied:
확인하여 4bd664b437ab1272955d0e5eeb76bb5f4eece641 에 반영하였습니다:)

귀한시간 내서 코드리뷰 해주셔서 감사합니다.
도율님 덕분에 많은 공부와 생각을 하게되었습니다!

ReadnThink committed:
refector : refectoring with clean code

- Sse로직을 SseSender를 사용해 처리하였습니다.
  - Sse를 전송하는 로직이 자주 사용되어 SendSse함수를 만들어
    재사용하도록 하였습니다.
- 실시간 매칭시 Crew를 만드는 것의 코드가 너무 길어져 빌더패턴을
  적용하지 않고 Static 메서드를 커스텀하여 Crew를 생성하였습니다.
  - 빌더패턴을 사용하는 이유는 로직이 변경되어도 객체 생성시 필드값을 자유롭게 설정하는
    것이였는데, 랜덤매칭시 Crew를 만들때는 인자가 제한적일것이라고
    판단하여 Static 메서드를 사용했습니다.
- 배열보다는 컬렉션을 사용하여 안전한 코드를 작성하였습니다.

ReadnThink commented:
4bd664b437ab1272955d0e5eeb76bb5f4eece641 커밋에는 아직 Redis에 User를 저장하고 있지 않습니다.
이후 리펙토링에서 변경할 예정입니다.

또한 전체적으로 메서드에 중복되는 인자를 감싸는 객체를 만들어
유지보수에 유연하고, 가독성 있는 코드를 작성할 예정입니다.

ReadnThink committed:
refector : Matching Cancle refectoring (HashMap -> RedisSet)

- 매칭 취소로직 HashMap -> RedisSet으로 변경
- 기존 불필요한 코드 정리

ReadnThink committed:
주석 및 들여쓰기 제거

doxxx93 commented:
많이 개선이 된 거 같군요

ReadnThink commented:
> 많이 개선이 된 거 같군요

도율님 덕분입니다ㅎㅎ
코드리뷰 같이 해주셔서 감사합니다:)
